src/ProjectFileIO.cpp

```diff
@@ -1895,7 +1895,13 @@ bool ProjectFileIO::WriteDoc(const char *table,
 {
    auto db = DB();
 
-   TransactionScope transaction(mProject, "UpdateProject");
+   // This method is called:
+   // - periodically when recording, in the AudioIO Thread (i.e. DrainBuffers)
+   // - when keystroking a label,    in the Main    Thread
```

Comment:
Comment should really say, it's called from the main thread whenever there is any editing operation.  Most of those are disabled during play and recording but addition of labels is one example that is not (but that's not the exhaustive list of cases).